kyuubi-events/src/main/scala/org/apache/kyuubi/events/handler/ElasticSearchLoggingEventHandler.scala

```diff
@@ -0,0 +1,137 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.events.handler
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.module.scala.DefaultScalaModule
+import com.sksamuel.elastic4s.{ElasticClient, ElasticProperties, RequestFailure}
+import com.sksamuel.elastic4s.ElasticApi.indexInto
+import com.sksamuel.elastic4s.ElasticDsl._
+import com.sksamuel.elastic4s.http.JavaClient
+import com.sksamuel.elastic4s.requests.common.RefreshPolicy
+import com.sksamuel.elastic4s.requests.mappings.MappingDefinition
+import com.sksamuel.elastic4s.requests.mappings.dynamictemplate.DynamicMapping
+import org.apache.commons.lang3.StringUtils
+import org.apache.hadoop.util.Preconditions.checkArgument
+import org.apache.http.auth.{AuthScope, UsernamePasswordCredentials}
+import org.apache.http.impl.client.BasicCredentialsProvider
+import org.apache.http.impl.nio.client.HttpAsyncClientBuilder
+
+import org.apache.kyuubi.Logging
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.config.KyuubiConf._
+import org.apache.kyuubi.events.KyuubiEvent
+import org.apache.kyuubi.events.handler.ElasticSearchLoggingEventHandler.getElasticClient
+
+/**
+ * This event logger logs events to ElasticSearch.
+ */
+class ElasticSearchLoggingEventHandler(
+    indexId: String,
+    serverUrl: String,
+    username: Option[String] = None,
+    password: Option[String] = None,
+    kyuubiConf: KyuubiConf) extends EventHandler[KyuubiEvent] with Logging {
+
+  private val isAutoCreateIndex: Boolean =
+    kyuubiConf.get(SERVER_EVENT_ELASTICSEARCH_INDEX_AUTOCREATE_ENABLED)
+
+  private val esClient: ElasticClient = getElasticClient(serverUrl, username, password)
+
+  private def checkIndexExisted: String => Boolean = (indexId: String) => {
+    checkArgument(
+      StringUtils.isNotBlank(indexId),
+      "index name must be configured for ElasticSearchEventHandler, please ensure %s is set",
+      SERVER_EVENT_ELASTICSEARCH_INDEX.key)
+    // check
+    esClient.execute {
+      indexExists(indexId)
+    }.await.result.isExists
+  }
+
+  private def checkAndEnsureIndex(indexId: String): Unit = {
+    if (!checkIndexExisted(indexId) && isAutoCreateIndex) {
+      esClient.execute {
+        createIndex(indexId)
+          .mapping(MappingDefinition(dynamic = Some(DynamicMapping.Dynamic)))
+      }.await
+      if (!checkIndexExisted(indexId)) {
+        throwExceptionIndexNotfound(indexId)
+      }
+    } else {
+      throwExceptionIndexNotfound(indexId)
+    }
+  }
+
+  checkAndEnsureIndex(indexId)
+
+  private val objectMapper = new ObjectMapper().registerModule(DefaultScalaModule)
+
+  override def apply(event: KyuubiEvent): Unit = {
+    val fields = {
```

Comment:
This does not seem to be a good solution that serializes the event into a json string and then deserializes it again into a map.  `elastic4s` provides the `indexInto(indexId).doc(XXX)` method to directly index json string.

Reply:
Tried, but not working. Due to the nested fields that I have to serialize-deserialize-map again. Tried to implement a method for generating flat json, but not serializing properly.

Reply:
> but not serializing properly.

Are there any specific errors?